src/Microsoft.DotNet.Wpf/src/PresentationFramework/Microsoft/Win32/OpenFolderDialog.cs

```diff
@@ -0,0 +1,215 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+//
+// 
+// Description:
+//              OpenFolderDialog is a sealed class derived from FileDialog that
+//              sets the pick folders option. The class does not support the legacy
+//              SHBrowseForFolder dialog.
+
+namespace Microsoft.Win32
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Security;
+    using System.Windows;
+
+    using MS.Internal.AppModel;
+    using MS.Internal.Interop;
+    using MS.Internal.PresentationFramework;
+    using MS.Win32;
+
+    /// <summary>
+    ///  Represents a common dialog box that allows the user to open one or more file(s). 
+    ///  This class cannot be inherited.
+    /// </summary>
+    public sealed class OpenFolderDialog : FileDialog
```

Comment:
Can this be a [CommonDialog](http://msdn.microsoft.com/library/Microsoft.Win32.CommonDialog) rather than a `FileDialog`? Making it a `FileDialog` gives it APIs that don't make sense for a folder, such as `CheckFileExists`, `DefaultExt`, `SafeFileName`, etc.

Reply:
CommonDialog has no fields and especially no validations, they're all in FileDialog. So either we copy everything except the few APIs you mentioned (which people have criticized on v1), or we just accept the fact that some APIs are irrelevant. Or we make a new base class for both, but that's a deeper change.

FYI: The Windows API Code Pack-Shell also has a FileDialog which becomes a Folder Dialog by setting a property to true. It also has SafeFileName (which will be the folder path then).

Reply:
We shouldn't have API that's known to never work though. We want great documentation, but at the same time we want to design easy APIs that don't require outside reading. Basic OOP design is that a subclass add functionality, not remove it.

What makes it difficult is that we also use subclassing for sharing implementations. If the `FileDialog` code is stable maybe a copy of the relevant parts of it into OpenFolderDialog is OK, and maybe as a separate PR when it's important enough. I think there are also places in code where common functionality has been put into static helpers to provide shared code for cases like this.